Add decomp rule to assert_tensor_metadata for BatchedTensors

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #163008


Whenever there is device move, export introduces assert_tensor_metadata aten operator to make sure to guard for device specialization. This aten op didn't work with Vmap because we didn't register explicit decomp rule saying we just skip BatchedTensor and call it on underlying tensor 

Differential Revision: [D82483979](https://our.internmc.facebook.com/intern/diff/D82483979)

tugsbayasgalan committed:
Fix vmap fallback logic when there is op that has 0 return

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82483979).

tugsbayasgalan committed:
Update on "Fix vmap fallback logic when there is op that has 0 return"



Whenever there is device move, export introduces assert_tensor_metadata aten operator to make sure to guard for device specialization. This aten op didn't work with Vmap fallback logic because there is an assertion that says return num should be more than 0 which seems overly strict. So in this PR, I just remove the restriction. 

Differential Revision: D82483979

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82483979).

tugsbayasgalan commented:
@pytorchbot label "topic: not user facing"

zou3519 reviewed (changes requested):
this is not the right fix, please don't merge it as-is. I'll be back to read this tomorrow

huydhn commented:
> this is not the right fix, please don't merge it as-is. I'll be back to read this tomorrow

Thank @zou3519 for looking into this!  Here is the issue FYI https://github.com/pytorch/pytorch/issues/162838

tugsbayasgalan committed:
Update on "Fix vmap fallback logic when there is op that has 0 return"



Whenever there is device move, export introduces assert_tensor_metadata aten operator to make sure to guard for device specialization. This aten op didn't work with Vmap fallback logic because there is an assertion that says return num should be more than 0 which seems overly strict. So in this PR, I just remove the restriction. 

Differential Revision: [D82483979](https://our.internmc.facebook.com/intern/diff/D82483979)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82483979).

tugsbayasgalan committed:
Update on "Add decomp rule to assert_tensor_metadata for BatchedTensors "



Whenever there is device move, export introduces assert_tensor_metadata aten operator to make sure to guard for device specialization. This aten op didn't work with Vmap because we didn't register explicit decomp rule saying we just skip BatchedTensor and call it on underlying tensor 

Differential Revision: [D82483979](https://our.internmc.facebook.com/intern/diff/D82483979)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D82483979).

tugsbayasgalan commented:
@pytorchbot merge -f "Landed internally"

zou3519 commented:
Does re-requesting the review dismiss the "request changes"?

huydhn commented:
@zou3519 This was forced merge, I should have revoked my previous approval.  Let me know if this looks ok or should it be reverted?

huydhn commented:
@pytorchbot --help

huydhn commented:
@pytorchbot cherry-pick --onto release/2.9 -fixes https://github.com/pytorch/pytorch/issues/162838 -c regression

huydhn commented:
@pytorchbot cherry-pick --onto release/2.9 --fixes https://github.com/pytorch/pytorch/issues/162838 -c regression
